File: examples/cookbook/07-multilang/check.lua
-- Cookbook 07 — assert the Lua face of the shared `journal` library.
-- Same library as check.py asserts for Python; compare the two side by side.

local journal = require("journal")

-- snake_case style: EVERY name reshaped for Lua, classes included.
local e = journal.entry("day 1", "welded")
assert(e:render_line() == "day 1: welded")

local nb = journal.notebook()
nb:add_entry(e)
nb:add_entry(journal.make_entry("day 2"))
assert(nb:entry_count() == 2)

-- Per-language weld_as: Lua spells renderAll `as_string` (Python: `to_text`).
assert(nb:as_string() == "day 1: welded\nday 2: (empty)\n")
assert(nb.to_text == nil)

-- The Lua-flavored save_to (mark::only(lua)) takes a writer callback.
local lines = {}
nb:save_to(function(line) table.insert(lines, line) end)
assert(#lines == 2)
assert(lines[1] == "day 1: welded")
assert(lines[2] == "day 2: (empty)")

assert(journal.format_version == 1)

-- Lua has no runtime docstrings; the LuaCATS stub carries them instead.
local stub_path = os.getenv("JOURNAL_LUACATS")
local f = assert(io.open(stub_path, "r"))
local stub = f:read("a")
f:close()
assert(stub:find("---@meta", 1, true))
assert(stub:find("journal.notebook", 1, true))    -- styled class name
assert(stub:find("save_to", 1, true))             -- the lua-only flavor is typed
assert(stub:find("dated journal entry", 1, true)) -- the doc text survives

print("cookbook 07-multilang (lua): OK")

File: examples/cookbook/07-multilang/check.py
"""Cookbook 07 — assert the Python face of the shared `journal` library.

Same library as check.lua asserts for Lua; compare the two side by side.
"""

import io
import os

import journal


def main() -> None:
    assert "journalling library" in journal.__doc__

    # PEP 8 style: camelCase C++ reshaped, classes stay PascalCase.
    e = journal.Entry("day 1", "welded")
    assert e.render_line() == "day 1: welded"
    assert not hasattr(e, "renderLine")

    nb = journal.Notebook()
    nb.add_entry(e)
    nb.add_entry(journal.make_entry(title="day 2"))
    assert nb.entry_count() == 2

    # Per-language weld_as: Python spells renderAll `to_text` (Lua: `as_string`).
    assert nb.to_text() == "day 1: welded\nday 2: (empty)\n"
    assert not hasattr(nb, "as_string")

    # The Python-flavored save_to (mark::only(py)) takes a file-like object —
    # anything with .write, e.g. io.StringIO or an open() handle.
    sink = io.StringIO()
    nb.save_to(sink)
    assert sink.getvalue() == nb.to_text()

    # Docstrings arrive Google-style (welder's doc/returns annotations).
    assert "dated journal entry" in journal.Entry.__doc__
    assert "Returns:" in journal.make_entry.__doc__

    # The style reshapes variables too: FORMAT_VERSION -> format_version.
    assert journal.format_version == 1

    # The .pyi stub (nanobind's bundled stubgen) carries the same surface.
    stub_path = os.environ["JOURNAL_PYI"]
    with open(stub_path, encoding="utf-8") as f:
        stub = f.read()
    assert "class Notebook:" in stub
    assert "def save_to(" in stub
    assert "def to_text(" in stub

    print("cookbook 07-multilang (python): OK")


if __name__ == "__main__":
    main()